add User, Manager Repository and AuthenticationMan

anitTer1789 opened the pull request:
added User in Model 
added Manager Repository in Repository
added AuthenticationManager and AuthenticationManagerImpl in Service

anitTer1789 committed:
add User, Manager Repository and AuthenticationMan

Adikso commented on src/main/java/net/io/kino/model/User.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.Optional;
+
+public class User {
```
Żeby użyć JpaRepository wraz z User to User musi mieć adnotacje `@Entity`. Bez tego aplikacja się nie uruchomi, bo dostaniesz błąd:
`Caused by: java.lang.IllegalArgumentException: Not a managed type: class net.io.kino.model.User`

Dodatkowo field z identyfikatorem musi być oznaczony adnotacją, żeby nie dostać błędu:
`Caused by: org.hibernate.AnnotationException: No identifier specified for entity: net.io.kino.model.User` 
oraz musisz ręcznie ustalić nazwę tabeli na np. "users", ponieważ nazwa "user" jest zarezerwowana w mssql. Obie te rzeczy są zrobione w Order.java

anitTer1789 replied:
Jasne, to się szybko doda.

Adikso commented on src/main/java/net/io/kino/model/User.java:
```diff
@@ -0,0 +1,33 @@
+public class User {
+    Long ID;
+    String password;
+    boolean status;
```
Gdy już tworzymy gettery i settery to raczej staramy się ograniczyć widoczność fieldów. Zmień może ich widoczność na private

Adikso replied:
Przydałaby się też jakaś nazwa użytkownika

anitTer1789 replied:
O, z tymi polami to faktycznie mój błąd. Nie zauważyłam. Nazwa użytkownika początkowo była, potem została zmieniona na ID. Ale to w takim razie dodam jeszcze jedno pole z nazwą.

Adikso commented on src/main/java/net/io/kino/service/impl/AuthenticationManagerImpl.java:
```diff
@@ -0,0 +1,80 @@
+    @Override
+    public boolean login(User user) throws Exception {
+        if(authenticate(user)) {
+            user.setStatus(true);
```
Nie do końca rozumiem co reprezentuje field status, zwłaszcza, że zapisywany jest do bazy danych. Czy ostatnie logowanie się udało? Ale gdyby już chcieć zapisywać informacje o ostatnim logowaniu w bazie to przydałoby się też jakieś ip.

anitTer1789 replied:
Pole status miało oznaczyć, czy obecnie użytkownik jest zalogowany. Ale przy głębszym namyśle to faktycznie nie ma sensu, z uwagi na przekazywany token. Zrezygnujemy z tego.

Adikso commented on src/main/java/net/io/kino/service/impl/AuthenticationManagerImpl.java:
```diff
@@ -0,0 +1,80 @@
+    // check if hash of a given password corresponds with a password in database
+    @Override
+    public boolean authenticate(User user) throws Exception {
+        User found = managers.getOne(user.getID());
```
Gdy przyjmujesz do metody obiekt modelu to raczej zakładasz w tym momencie, że z frontendu dostajesz jsona, który deserializuje się do klasy User. Żeby ta linijka działała to frontend musiałby znać id użytkownika, ale go nie zna, ponieważ z reguły użytkownik nie podaje swojego id liczbowego przy logowaniu.

Warto rozważyć czy metody login/authenticate nie mogłyby po prostu przyjmować dwóch stringów login i password.

anitTer1789 replied:
Jasne, przyjmowanie login i password wydaje się jak najbardziej sensowe. Trzymaliśmy się tutaj naszej prezentacji i tam przekazywaliśmy do metody obiekt klasy User. Ale to w takim razie to zmienię.

Adikso commented on src/main/java/net/io/kino/repository/ManagersRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ManagersRepository extends JpaRepository<User, Long> {
```
Gdy jest User a nie Manager to UserRepository bardziej pasuje

anitTer1789 replied:
Myślę, że zmienię nazwę klasy z User na Manager, bo to chyba będzie sensowniejsze z uwagi na poprzedni komentarz.

Adikso reviewed (changes requested):
Wasz moduł powinien pozwalać też na sprawdzanie czy użytkownik jest zalogowany gdy wysyła jakieś żądania, więc pewnie logowanie powinno zwracać jakiś token i później ten token jest wysyłany z każdym żądaniem i weryfikowany. Ale jak to powinno wyglądać to dogadajcie się z @4Kaze (Jacek), bo on będzie tego używał w rest

th7nder commented on src/main/java/net/io/kino/service/AuthenticationManager.java:
```diff
@@ -0,0 +1,10 @@
+
+
+public interface AuthenticationManager {
+    boolean login(User user) throws Exception;
```
Rzucanie tak ogólniej klasy wyjątku jak Exception nie jest dobrym pomysłem, tak na prawde nie wiadomo co się stało.

anitTer1789 replied:
Wiesz co, miałam dokładnie to samo zdanie, ale z jakiegoś powodu metoda hash miała problemy z rzucaniem czymś innym. Ale spokojnie, to się zmieni

th7nder commented on src/main/java/net/io/kino/service/impl/AuthenticationManagerImpl.java:
```diff
@@ -0,0 +1,80 @@
+        String stored = found.getPassword();
+
+        String[] saltAndHash = stored.split("\\$");
+        if (saltAndHash.length != 2) {
```
Jeśli pobieramy dane z bazy to chyba zakładamy, że są już poprawne.

anitTer1789 replied:
Wiesz, przezorny zawsze ubezpieczony. Mogłaby się pojawić sytuacja, że ktoś tam coś ręcznie wpisał. Ja zawsze wolałam wszystko dwa razy zabezpieczać. Jeśli uważacie, że to za dużo, to można to wywalić, ale jak dla mnie to zawsze lepiej się upewnić, że pobrane dane są poprawne

anitTer1789 committed:
deleted User.java

th7nder commented on src/main/java/net/io/kino/service/impl/AuthenticationManagerImpl.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    // compute salted PBKDF2 hash of given password
+    public static String getSaltedHash(String password) throws Exception {
```
throws Exception jest tylko wymagane, jeśli rzucany wyjątek jest jest unchecked.

anitTer1789 replied:
Jak napisałam w poprzednim komentarzu, rzucane wyjątki poprawię. Po prostu podczas pisania pojawiły się małe konflikty, jak to często bywa.

anitTer1789 committed:
added Manager.java

added field "username". Changed fields to private. Fields status and ID are currently deleted

anitTer1789 committed:
changes made in ManagersRepository

anitTer1789 committed:
changed AuthenticationManger

changes in methods declarations

anitTer1789 committed:
changed AuthenticationManagerImpl

changed methods according to the declarations in interface
todo - passing login tokens and logout method

anitTer1789 committed:
Changed thrown Exceptions

Switched Exception for InvalidKeySpecException and NoSuchAlgorithmException

Adikso commented:
O co chodzi z tymi zmianami w ostatnim commicie w innych modułach i dodaniu klas dotyczących loggera, które importują klasy, które nie istnieją na tym branchu?

anitTer1789 committed:
deleted ActionLogger

file was added by accident

anitTer1789 committed:
deleted LogWriter

file was added by accident

anitTer1789 committed:
deleted ReservationService

file was added by accident

anitTer1789 committed:
deleted CinemaActionImpl

file was added by accident

anitTer1789 committed:
delted LocalDestinationLogWriterImpl

file was added by accident

anitTer1789 committed:
delted ReservationServiceImpl

file was added by accident

th7nder commented on src/main/java/net/io/kino/service/impl/AuthenticationManagerImpl.java:
```diff
@@ -0,0 +1,68 @@
+
+    // compute salted PBKDF2 hash of given password
+    public static String getSaltedHash(String password) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        byte[] salt = SecureRandom.getInstance("SHA1PRNG").generateSeed(saltLen);
```
Jeśli chodzi o te wyjątki NoSuchAlgorithmException i InvalidKeySpecException, to chyba bym je tutaj złapał po prostu i rzucił jakiś IllegalState w zamian. 
Używając tych "zarażamy" dalszy kod programu, a i tak nikt z tym nic nie zrobi. Użytkownik tego API nie ma wpływu na to co się dzieje w tej metodzie, więc nie ma temu jak zapobiec.

anitTer1789 replied:
Ok, to tak zrobię

Adikso commented:
A to weryfikowanie tokenów i innych rzeczy to będzie w osobnym PR?

Adikso commented on src/main/java/net/io/kino/service/impl/AuthenticationManagerImpl.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Override
+    public boolean logout(Manager user) {
+        return false;
```
Podejrzewam, że ta metoda będzie zmieniona, ale jako sama deklaracja to co oznacza zwracany boolean z logout? Może się wylogowywanie nie udać? Jak się nie uda to co mogłoby to znaczyć i co z tym możemy zrobić?

anitTer1789 replied:
Tam wpisałam w komentarzu, że to do zmian. Pomysł był taki, żeby raczej po prostu dać sygnał, że wylogowywanie się powiodło, gdyby któryś z modułów tego potrzebował (mięliśmy prezentacje przed frontendem, także woleliśmy to uwzględnić). Myślę, że jednak zmienię to na void, ale jeszcze muszę pomyśleć, jak ta metoda powinna wyglądać.

th7nder reviewed (dismissed):
Pliki ReservationService i ReservationServiceImpl zostały usunięte, czy tak powinno być?

anitTer1789 commented:
> A to weryfikowanie tokenów i innych rzeczy to będzie w osobnym PR?

Na razie się zastanawiam, jak to zrobić. Być to będzie, ale jeszcze pomyślę gdzie.

anitTer1789 commented:
> Pliki ReservationService i ReservationServiceImpl zostały usunięte, czy tak powinno być?

Z jakiegoś powodu te pliki zostały przesłane wraz z innymi w ostatnim commicie i pokazało, że są zmiany względem tego, co było. Prawdopodobnie to była jakaś starsza wersja, która znalazła się u mnie na komputerze. Jak powiedziałam, nie miałam zamiaru tego wstawić u siebie

Adikso committed:
Revert deletion of ReservationService

Adikso committed:
Revert deletion of ReservationService v2

Adikso commented:
Żebyś się nie bawiła z tokenami do resta to możemy zrobić to z Basic Auth
https://gist.github.com/Adikso/7799ba3f56b296bcf5876cb7dd69a64e

Tak się ogólnie nie robi w prawdziwości, ale well

Tylko tutaj trzeba to jeszcze przerobić, bo żeby to hasło przyjęło to musi być informacja o password encoderze do hasła. Czyli np w naszym przypadku "{noop}" przed hashem hasła. Możesz zrobić "hacka" na to i do tego zapytania z pliku wyżej do password concatować to z przodu albo możesz ucinać przy sprawdzaniu hasha czy coś

anitTer1789 commented:
> Żebyś się nie bawiła z tokenami do resta to możemy zrobić to z Basic Auth
> https://gist.github.com/Adikso/7799ba3f56b296bcf5876cb7dd69a64e
> 
> Tak się ogólnie nie robi w prawdziwości, ale well
> 
> Tylko tutaj trzeba to jeszcze przerobić, bo żeby to hasło przyjęło to musi być informacja o password encoderze do hasła. Czyli np w naszym przypadku "{noop}" przed hashem hasła. Możesz zrobić "hacka" na to i do tego zapytania z pliku wyżej do password concatować to z przodu albo możesz ucinać przy sprawdzaniu hasha czy coś

Możemy zrobić to z Basic Auth, bo domyślam się, że to pójdzie szybciej i sprawniej. Obecnie większy problem jest raczej z wylogowaniem użytkownika, bo w sumie tak naprawdę to nie wiem, ile ja mogę tutaj zdziałać czystą javą. Teoretycznie jedyne, co mogę to unieważnić token, ale jak tak czytam, to nie widzę na to żadnego sensownego sposobu.

Adikso commented:
Z Basic Auth to problemu nie ma bo cały czas wysyłamy login i hasło zakodowane w nagłówku więc nie ma sesji (nie ma logout)

anitTer1789 commented:
Ok, no to faktycznie tak zrobię.

anitTer1789 committed:
Update AuthenticationManager

deleted thrown exceptions

anitTer1789 committed:
updated AuthenticationManagerImpl

using try-catch blocks instead of throwing exceptions

anitTer1789 committed:
added SecurityConfiguration

anitTer1789 committed:
Update AuthenticationManagerImpl.java

Adikso commented:
Brakuje dependency do spring security w gradle

anitTer1789 commented:
O, dzięki za zwrócenie uwagi. Już dodaję

anitTer1789 committed:
added spring security dependency

Adikso commented:
Musisz chyba jednak zrobić inaczej to logowanie:
https://www.baeldung.com/spring-security-authentication-provider
Patrz tylko na punkt 2. i 4.
Zamiast `shouldAuthenticateAgainstThirdPartySystem` to twoja metoda login

Bo to co aktualnie jest to w sumie tylko porównuje stringi dwa a nie sprawdza hash (czyli wychodzi trochę jakbyś hasła w plaintext trzymała w bazie)

Adikso commented:
I to chyba wszystko będzie (?)

anitTer1789 committed:
changed login method declaration

anitTer1789 committed:
changed login method

anitTer1789 commented:
Chwila, podmieniłam metodę login, żeby działała analogicznie jak tamta metoda authenticate, bo robienie nowej metody, która wywoływałaby login to nie miałoby raczej sensu, bo login tylko wywołuje moje authenticate. A jeśli chodzi o przechowywanie hashy, to tam przecież porównuję dwa hasze. Czytałam, że tamten generowany przeze mnie PBKDF2 hash jest bezpieczny do przechowywania w bazie.

Adikso commented:
Trochę o coś innego chodziło. Klasa z linku `CustomAuthenticationProvider` to powinna być właśnie osobna klasa i tam wykonanie tego `authenticate`. Dokładnie tak jak w przykładzie  tylko inna metoda użyta.

I trzeba też w security configuration ustawić, żeby to było używane
```java
    @Autowired
    private CustomAuthenticationProvider authProvider;
 
    @Override
    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
        auth.authenticationProvider(authProvider);
    }
``` 
zamiast tamtych zapytań do bazy w metodzie `configure`

Z tym bezpieczeństwem to chodzi mi o to, że to zapytanie co było w SecurityConfiguration to pobierało z bazy hash i porównywało go jak tekst z przesyłanym hasłem. Wtedy klient musiałby wysyłać hash hasła, a to niszczyło by cały sens używania hashy. Przykładowo jakby ktoś zrobił atak na system i wykradł bazę danych to gdy wysyłamy hashe to mógłby od razu się nimi zalogować. Gdy wysyłamy hasło w plaintext i dopiero api to hashuje i porównuje to atakujący musiałby najpierw złamać te hashe

anitTer1789 commented:
A, rozumiem już. Ok, to dodam klasę i podmienię tamte ustawienia

anitTer1789 committed:
updated SecurityConfiguration, added CustomAuthProvider

anitTer1789 committed:
changed login declaration

anitTer1789 committed:
changed login method

Adikso commented:
Brakuje adnotacji `@Component` do `CustomAuthenticationProvider`. Bez tego się nie odpali

anitTer1789 committed:
added @Component annotation

Adikso commented on src/main/java/net/io/kino/security/CustomAuthenticationProvider.java:
```diff
@@ -0,0 +1,40 @@
+
+public class CustomAuthenticationProvider implements AuthenticationProvider {
+
+    private AuthenticationManager auth;
```
Brakuje `@Autowired`

anitTer1789 replied:
Już jest

anitTer1789 committed:
added component annotation

anitTer1789 committed:
added Autowired annotation

Adikso committed:
Fix login service

Adikso committed:
Fix formatting

Adikso commented:
Pierwsza na mecie
